### ISSUE TITLE
Would like option to "init" a mix in offline mode and just use local RPMs

### ISSUE THREAD

**Issue by bktan8** [clearlinux/mixer-tools#283]
Today, "mixer init" requires a network connection to the upstream server.
It would be nice to initiate a mix in completely offline mode and just use local RPMs.

**Comment by cmarcelo** [comment]
Typically a "mix" is based on some existing version of Clear Linux (or of other mix). For init, you need to have only available somehow so it can bootstrap information (e.g. bundle information, and the upstreamversion).

It probably would help understand your larger goal to see what is the best way to solve your issue. E.g. do you want to create a new mix without any content from Clear Linux or other mix, just your RPMs? Or do you want to create a mix of Clear but just not hit the network?

**Comment by bktan8** [comment]
@cmarcelo - we have a customer interested in the ability to mix completely offline.  

I was able to archive the upstream packages and bundle definitions and move them over to "local" and completely mix offline ONCE I get pass the "init" stage which still requires a network connection.  The global flag "--offline" doesn't seem to apply to "init".  

I'm looking to see if we can make "init" create boiler plates that can be used to create a mix without the network.

**Comment by cmarcelo** [comment]
@bktan8 Thanks for the explanation. I'm not sure about whether will this mix get updates... i.e. if it uses a package "glibc" in the Mix, and Clear releases a new version -- what is the plan to get this new version in the mix?

The mixer tool has some codepaths to not use the upstream (which are used to create Clear Linux itself), it seems like that "mode" would be useful in this case. Since "init" is not used to create Clear Linux itself, it always assumed an upstream existed.

**Comment by ahkok** [comment]
Effectively the question would be a request to be able to have the `master` be changeable to a different server so mixes can be mixed. I don't see any reason to not support it, after all, the only thing needed would be to change `cdn.download.clearlinux.org/update` into a configuration variable. And then people can make it point to `http://127.0.0.1/` or even `file:///var/bktan8`. This would be accomplished by setting up `NAT` to transparently proxy it as well, just makes it simpler for the user.

The caveat is that the new `master` needs to have the expected folder organization as is expected, and this would need documentation, I think. But I don't think this request is misplaced.

**Comment by cmarcelo** [comment]
> But I don't think this request is misplaced.

@ahkok agreed. We already have some plumbing for this (i.e. supporting mixing from an upstream other than Clear Linux itself), but might need some tweaks. That is slightly different than I was thinking was being asked, but might be enough for @bktan8 use case.

**Comment by tmarcu** [comment]
The option to set your own server or backup servers etc currently exists with the `mixer repo add|remove` command. I am worried about what @cmarcelo points out, if all they are doing is creating a mirror of upstream Clear, they will have to re-pull all the RPMs again for that new upstream version each time it's bumped - which they should do so they don't stay pinned to an old release. At that point it's no different than just letting mixer pull what it needs directly.

**Comment by matthewrsj** [comment]
> the only thing needed would be to change cdn.download.clearlinux.org/update into a configuration variable.

This can already be set in the builder.conf. I've used it for my testing many times to point to an internal server when the proxy or CDN is being too slow.

**Comment by kevincwells** [comment]
>> the only thing needed would be to change cdn.download.clearlinux.org/update into a configuration variable.

>This can already be set in the builder.conf. I've used it for my testing many times to point to an internal server when the proxy or CDN is being too slow.

This is true for everything _except_ the bundle definitions themselves, for which we pull a release tarball from [github](https://github.com/clearlinux/clr-bundles/releases)

**Comment by matthewrsj** [comment]
> we pull a release tarball from github

Ah, yes this should be configurable as well.

**Comment by tmarcu** [comment]
The bundle definitions should be updated to configurable too 👍  
I understand client machines needing to be off-line or restricted to a subnet pulling from a custom server, but why does the mix server itself have to be isolated from the network? <- in the case that you are NOT providing every single RPM yourself, and have to go download them from download.clearlinux.org still anyway before you can build.

**Comment by seanvk** [comment]
A downstream customer  with many Clear based clients will already be mixing their own releases and setting builder.conf to their update server.  So having a configurable bulder.conf for offline mixing is a doable thing.

**Comment by kevincwells** [comment]
> I was able to archive the upstream packages and bundle definitions and move them over to "local" and completely mix offline ONCE I get pass the "init" stage which still requires a network connection. The global flag "--offline" doesn't seem to apply to "init".

Just to be completely clear, `--offline` _does_ "work for `mixer init`", but it's probably a little unclear what it's actually doing. Let me try to clarify a little what's going on.

`mixer init` does two things relevant to this conversation:
1. Pulls down the upstream package list and bundle definition files, and puts them in `upstream-bundles/`.
1. Makes an internal call to `AddBundles` for the four default bundles: os-core, os-core-update, bootloader, and kernel-native.

What `--offline` actually does is make the upstream fetch (that pulls down the package list and bundle definitions) abort immediately, essentially skipping the first step.

That second step doesn't care where those four bundles come from (upstream or local), but it needs to find them somewhere. If you skipped fetching them in the first step, **and you haven't done anything else**, you'll get an error in step 2 saying it can't find the bundles.

Thus, if you want `--offline` to work for `mixer init`, you'll need to copy your local versions of those four bundles into `local-bundles` **before** you call `mixer init --offline`.

**Comment by kevincwells** [comment]
**A very important thing to note**, though, is that even if you have local copies of the bundle definitions cached as @bktan8 has described, this **does not actually let you "completely mix offline"**.  As mentioned above, the `--offline` tag stops `mixer` from trying to fetch bundle definition content from upstream. **During an actual build, mixer still fetches all upstream package content from upstream via its `dnf` calls.**

**Comment by kevincwells** [comment]
@bktan8 -- Ah, yup! The only way to do a true full offline build is to mirror _every_ RPM from _every_ upstream bundle in your mix to your local dnf repo, like you just described. Not an easy task for a large mix with many bundles. Out of curiosity, how did you go about building the list of RPMs you needed to clone from upstream?

**Comment by bktan8** [comment]
@kevincwells - I just cloned everything with: 

           $ wget -A.rpm --no-verbose --no-parent --recursive --level=1 --no-directories      
           https://download.clearlinux.org/releases/22290/clear/x86_64/os/Packages/

Customer is also interested in scenario where they can archive different versions (in case upstream no longer exists) and go back and mix with.

**Comment by kevincwells** [comment]
Hey @bktan8 -- thanks for sharing your full workflow above. Let me offer some feedback:

Part of Step 1 and Step 4 cancel each other out. All `--all-upstream` does in Step 1 is add every single upstream bundle to your `mixbundles` list. It doesn't affect anything else (like which bundles are fetched to `upstream-bundles/`). You then take them all back out in Step 4, so you could skip this flag and leave off Step 4 altogether.

In Step 5, since you said you aren't actually editing the bundles and just copying them `local-bundles/`, you can do that all in one step: `mixer bundle edit --offline --suppress-editor os-core os-core-update kernel-native bootloader`.  (Technically, with your exact flow, you don't need the `--offline` here, because your `init` was already done without `--offline`, meaning the bundles were already cached once. `mixer` will not reach back out to the network to re-download the bundles if it already finds a local copy.)

**Comment by tmarcu** [comment]
From my offline conversation with @bktan8 it seems the underlying issue here is a need to have content persist forever, so that it is always accessible regardless of the state of our upstream content. Being able to build offline once a build is done can be easily achieved by enabling a cache, so mixes can be created again when you're on a plane or conference with no network. However, this issue points to being able to go back and always have access to the RPM content of any upstream release.

This can be done, but I would like to point out some very important considerations:

1.) You are **not** technically building offline according to this flow. You are downloading *.RPMs with wget (you need network for this still), and then adding them locally. This points to -> you are in need of a cache, or -> you don't trust Clear to provide content & Mixer to download the RPMs for you. Mixer will download _only_ the RPMs that you need to create your mix, which is likely a significantly smaller download. If you want full mix capabilities from local only, you **must download _all_ content for _every_ single upstream version you want to build against.**
This means you have to mirror many GB of data for every release of Clear (or ideally as often as possible) to stay up to date. You also have to maintain this mirror. This is a scenario groups in other geographies have had to do, because their connection is so unreliable/slow that it's unfeasible to work without a mirror; I don't believe this is the general case, though. Our Clear releases do not disappear, as seen from release 300 (first public release on download.clearlinux.org/update) still being available, so the persistence is there already. To re-state @kevincwells from offline, In Clear, we are the providers of *all* upstream content, and that content existing and being available, archived, is a fundamental requirement for our own OS and updates to work, not just for our derivatives.

2.) I believe this encourages bad behavior because it does not follow our expected build flow and BKMs for developing Clear - especially if it's not setup correctly. From historical experiences, this kind of flow paves the way to picking some upstream version, and building from it for ~6 months trying to get all custom content working and fixed. That mix is now several _hundred_ versions behind, and updating to latest which likely has a significant number of changes (critical/security fixes included) becomes very difficult. This has happened before.
Clear is a rolling release OS, and our ability to react to CVEs and push out fixes across the OS is a major selling point and philosophy we want to upkeep. **I really want to stress how much easier it is to always be up to date with the latest content, and push in incremental fixes as needed, than it is to try and catch up when you are behind.** 


Things we will need to enforce and encourage:
- Our Clear releases do not disappear. We ourselves have mirrors and backups. If Clear were to go down, our content does not instantly disappear in any scenario. And if it ever were the case that it went down for good, no derivative would continue using it as their base OS, they would switch entirely.
- Mirrors of our content need to be kept up to date and PREFERABLY only used as last resort backups, or if the network conditions make it so it's more efficient to build from the mirror.
- Follow BKMs and conventional workflows as much as possible, because it makes future integration much easier.

Going off that last point, I would like to revise your workaround for the time being @bktan8 to:
1.) Mirror the RPMs _and_ the repodata directory to a custom server, which will host the repo
2.) Clone the clr-bundles/ repo with the tag of the upstream version of the repo
2.) Now you can simply edit the dnf conf (.yum-mix.conf) with `mixer repo` command so the upstream repo points to that server location instead of `https://download.clearlinux.org/releases/$releasever/clear/x86_64/os/`
3.) ALL mixer usage and commands can now be performed as normal, and something like 

> mixer bundle add vim or --all-upstream && mixer build

 will add the bundles and pull RPMs from "upstream" which is actually the mirror server.
4.) Building and adding local content all remains the same.

*NOTE, that I do not encourage nor support this mode of operation unless the mirrored repo is being synced with our external releases often, and the upstreamversion file is updated with it so mixes are built with the most up-to-date content available. Otherwise you fall into the trap of never updating your upstream version and falling far behind in the stream.

I'll keep trying to think of how to make using the official stream the best option :) And a better way to enable offline caches for downstream OSV's.

**Comment by matthewrsj** [comment]
> During an actual build, mixer still fetches all upstream package content (RPMs) from upstream via its dnf calls.

You can remove remote repos by running `mixer repo remove <name>`. The names of the repos you have configured can be found with `mixer repo list`. So if you have all the RPMs local you can remove the default `clear` repo or `mixer repo set-url clear <new-url>` to a local repo url.

**Comment by tmarcu** [comment]
Closing as the propose solution is working.
